### DISCUSSION TITLE
Can not use `Self` in `AbstractContextManager` ?

### DISCUSSION THREAD

**Discussion by WSH032** [microsoft/pyright#7019] [Q&A] [answered]
Pyright complains that `Self` cannot be used in `AbstractContextManager`, but type inference works correctly.

Pylance does not complain about this, but type inference fails.

```py
from typing_extensions import Self, reveal_type
from contextlib import AbstractContextManager

# Pylance does not complain
# But cli pyright complains:
#     error: "Self" is not valid in this context (reportGeneralTypeIssues)
class Foo(AbstractContextManager[Self]):
    def __enter__(self) -> Self:
        return self
    def __exit__(self, *_) -> None:
        pass

class Bar(Foo):
    pass


with Foo() as foo:
    # pylance:
    #     foo: Foo[Unknown]
    # pyright:
    #     information: Type of "foo" is "Foo"
    reveal_type(foo)

with Bar() as bar:
    # Both pylance and pyright:
    #     bar: Bar
    reveal_type(bar)

```

---

But this will work well

```py
class Foo(AbstractContextManager["Foo"]):
    def __enter__(self) -> Self:
        return self
    def __exit__(self, *_) -> None:
        pass
```

---

pyright `v1.1.346`
pylance `v2023.12.103`
typing_extensions `v4.9.0`
python `3.10.8`

**Comment by erictraut** [comment, marked as answer]
Pylance is a few months behind in updating pyright. Pyright's current behavior here is correct. The use of `Self` is explicitly [forbidden as a base class in a class definition](https://typing.readthedocs.io/en/latest/spec/generics.html#valid-locations-for-self) by the typing spec (and by PEP 673, which introduced `Self` to the type system). I've been working to tighten up pyright's conformance with [the typing spec](https://typing.readthedocs.io/en/latest/spec/index.html) now that it has been formalized and [conformance tests](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) are under development.

**Reply by WSH032** [reply]
Thank you very much for your prompt response and clarification! 💖

I have only read the documentation of [typing](https://docs.python.org/3/library/typing.html?highlight=classvar#typing.Self), which did not directly mention `Self is explicitly forbidden as a base class in a class definition`. So I feel confused that there is inconsistency between pyright and pylance behavior.

Really appreciate the information you provided and am pleased to see pyright enhancing consistency with type specifications. I will follow the behavior of pyright.
